base/inc/CopCore/include/CopCore/Ranluxpp.h

```diff
@@ -144,6 +161,21 @@ class RanluxppDouble : public RanluxppEngineImpl<52> {
 
   __host__ __device__
   uint64_t IntRndm() { return this->NextRandomBits(); }
+
+  /// Branch a new RNG state, also advancing the current one.
+  __host__ __device__
+  RanluxppDouble Branch()
+  {
+    // Save the current state, will be used to branch a new RNG.
+    uint64_t oldState[9];
+    this->SaveState(oldState);
+    this->Advance();
+    // Copy and modify the new RNG state.
+    RanluxppDouble newRNG(*this);
+    newRNG.XORstate(oldState);
+    newRNG.Advance();
```

Comment:
I guess xor-ing two consecutive states branches in a cheap and reproducible way, but why do you need to advance the newRNG?

Reply:
Some RNGs need to be warmed up, not sure if this is the case for this particular RNG. However, I'd argue that one of the states could continue to be used without advancing as well, while the other could be branched and advanced.

Reply:
> [...] but why do you need to advance the newRNG?

Actually I've asked myself the same question while having a walk. I added this because the whole approach is modeled after MIXMAX, but they use an LCG to mangle only a single component whereas I think it's better to XOR the whole state (after all, the states should already be independent).

> However, I'd argue that one of the states could continue to be used without advancing as well, while the other could be branched and advanced.

No, we need to advance the original state to make sure that the next branching doesn't end up in the same state. But there's a different "reuse" possible that I've already been working on in the afternoon, stay tuned :wink:

Reply:
I ran some tests over the weekend and the errors in the tail worry me: [fig_CaloEDep_NoField.pdf](https://github.com/apt-sim/AdePT/files/6294438/fig_CaloEDep_NoField.pdf) This could mean that there's too much correlation between the states when only XORing. This is avoided by an immediate `Advance()` and the strong guarantees of RANLUX++. I'd prefer not to go this way, at least for now, until we understand if that's a real error or just statistical.

Reply:
OK, we can keep the Advance if it reduces the correlations. Anyway, I think we need to understand those in particular for the case without field. One could think of rounding effects, is there anything in float in any of the three cases (G4 native, G4HepEm, and AdePT-G4HepEm? The last two seem to be free of float for the energy deposition, but can be due also to intermediate calculations. I would first check that.

Reply:
Look, if I get correlations when deleting that line, how can it be `float`?

Reply:
But there are correlations compared to G4 results even in the initial TestEm3 version, at least this is how I read the original plots. I'm certainly not excluding some are coming from the xor.

Reply:
There are *errors* in the permille range, and there will be always be: even between versions of Geant4, you don't get exactly the same results. What worries me, is that they grow bigger.